### PR TITLE
[WIP] feat: support createRequire

### LINKED ISSUE #14228: feature request: createRequire detect
## Feature request

**What is the expected behavior?**

For code like below:

```js
import { createRequire as _createRequire } from 'module'
const __require = _createRequire(import.meta.url)
```

`__require` should be detected as an alias of CommonJS require and the imported module of `__require` should be the type of `javascript/auto` instead of `javascript/esm`.

**What is motivation or use case for adding/changing the behavior?**

This allows webpack to bundle this format of ESModule projects.

And [TypeScript is going to add Node-12 module emit](https://github.com/microsoft/TypeScript/pull/44501) which will generate code like above.

**How should this be implemented in your opinion?**

Detect import of `module` module and track the pattern of `const $requireAlias = $importedCreateRequire(import.meta.url)`.

**Are you willing to work on this yourself?**
yes but I'm not familiar with webpack internal.

### PR DESCRIPTION
**Pull request by Jack-Works**
close #14228

**What kind of change does this PR introduce?**

Support the following pattern as import CJS from ESM

```js
import { createRequire as _createRequire } from 'module'
const __require = _createRequire(import.meta.url)
```

**Did you add tests for your changes?** Not yet

**Does this PR introduce a breaking change?** No

**What needs to be documented once your changes are merged?** Not yet

### CODE REVIEW

File: lib/dependencies/HarmonyModulesPlugin.js [hunk 1]
```diff
@@ -103,6 +104,9 @@ class HarmonyModulesPlugin {
 						return;
 
 					new HarmonyDetectionParserPlugin(this.options).apply(parser);
+					new HarmonyNode12CreateRequireDependencyParserPlugin(
```
Comment (alexander-akait) [inline lib/dependencies/HarmonyModulesPlugin.js:107]: Not only related to node v12, should be renamed

**Review by alexander-akait — COMMENTED** [review]
Also I think we should provide option for register custom names, like we do for web workers https://webpack.js.org/guides/web-workers/

File: lib/dependencies/HarmonyModulesPlugin.js (hunk 1 above)
Comment (Jack-Works) [reply lib/dependencies/HarmonyModulesPlugin.js:107]: `createRequire` exported from `module` firstly comes from Node 12 API, I'm following the TypeScript naming (they choose to name those codegen targets as `Node12` and `NodeNext`)

**Comment by Jack-Works** [comment]
> Also I think we should provide option for register custom names, like we do for web workers https://webpack.js.org/guides/web-workers/

My intention was to provide support for the usage of Node `createRequire` API as a kind of `require` usage. It's very possible that different libraries authors will write different `const $NAME_HERE = createRequire(import.meta.url)`. My current design supports any top-level custom require function created by `createRequire`.

File: lib/dependencies/HarmonyModulesPlugin.js (hunk 1 above)
Comment (alexander-akait) [reply lib/dependencies/HarmonyModulesPlugin.js:107]: We don't have to follow the name in ts, I don't think we will have more `createRequire` dependencies

**Comment by Jack-Works** [comment]
I found this is not needed. NodeJS has a little part of CJS-ESM co-op support.

**Pull request closed without merging**